pkg/prometheus/promcfg.go

```diff
@@ -2318,8 +2320,25 @@ func (cg *ConfigGenerator) generateScrapeConfig(
 		cfg = append(cfg, yaml.MapItem{Key: "scrape_timeout", Value: *sc.Spec.ScrapeTimeout})
 	}
 
-	if sc.Spec.RelabelConfigs != nil {
+	if len(sc.Spec.RelabelConfigs) > 0 {
 		relabelings = append(relabelings, generateRelabelConfig(labeler.GetRelabelingConfigs(sc.TypeMeta, sc.ObjectMeta, sc.Spec.RelabelConfigs))...)
+	}
+
+	if shards != 1 {
+		alreadyHasShardLabel := false
+		for i, relabelConfig := range sc.Spec.RelabelConfigs {
+			if relabelConfig.Action == "hashmod" {
+				level.Debug(cg.logger).Log("msg", "found existing hashmod relabeling rule, skipping", "idx", i)
+				alreadyHasShardLabel = true
+				break
+			}
+		}
+		if !alreadyHasShardLabel {
+			relabelings = generateAddressShardingRelabelingRules(relabelings, shards)
```

Comment:
(suggestion) we could also have a new `generateAddressShardingRelabelingRulesIfMissing` function (sorry for the long name, feel free to use something shorter!). It would make code a bit more readable IMHO.